pytket/extensions/cutensornet/tnstate/ttn_gate.py

```diff
@@ -0,0 +1,376 @@
+        # p  -> the parent bond of the TTN node
+        # i  -> the input bond of the gate
+        # o  -> the output bond of the gate
+
+        node_bonds = [f"q{x}" for x in range(n_qbonds)] + ["p"]
+        result_bonds = node_bonds.copy()
+        node_bonds[target] = "i"  # Target bond must match with the gate input bond
+        result_bonds[target] = "o"  # After contraction it matches the output bond
+
+        # Contract
+        new_tensor = cq.contract(
+            node_tensor,
+            node_bonds,
+            gate_tensor,
+            ["o", "i"],
+            result_bonds,
+            options={"handle": self._lib.handle, "device_id": self._lib.device_id},
+            optimize={"path": [(0, 1)]},
+        )
+
+        # Update ``self.nodes``
+        # NOTE: Canonicalisation of the node does not change
+        self.nodes[path].tensor = new_tensor
+        return self
+
+    def _apply_2q_gate(self, q0: Qubit, q1: Qubit, gate: Op) -> TTNxGate:
+        """Applies the 2-qubit gate to the TTN.
+
+        Truncation is automatically applied according to the paremeters
+        in the ``Config`` object passed to this ``TTN``.
+        The TTN is converted to canonical form before truncating.
+
+        Args:
+            q0: The 0-th qubit the gate acts upon.
+            q1: The 1-st qubit the gate acts upon.
+            gate: The gate to be applied.
+
+        Returns:
+            ``self``, to allow for method chaining.
+        """
+        options = {"handle": self._lib.handle, "device_id": self._lib.device_id}
+
+        # Load the gate's unitary to the GPU memory
+        gate_unitary = gate.get_unitary().astype(dtype=self._cfg._complex_t, copy=False)
+        gate_tensor = cp.asarray(gate_unitary, dtype=self._cfg._complex_t)
+        # Reshape into a rank-4 tensor
+        gate_tensor = cp.reshape(gate_tensor, (2, 2, 2, 2))
+
+        (path_q0, bond_q0) = self.qubit_position[q0]
+        (path_q1, bond_q1) = self.qubit_position[q1]
+
+        # If the two qubits are in the same leaf node, contract the gate with it.
+        # There is no truncation needed.
+        if path_q0 == path_q1:
+            n_qbonds = (
+                len(self.nodes[path_q0].tensor.shape) - 1
+            )  # Total number of physical bonds in this node
+
+            # Glossary of bond IDs
+            # qX -> where X is the X-th physical bond (qubit) in the TTN node
+            # p  -> the parent bond of the TTN node
+            # i0 -> the input bond of the gate on q0
+            # o0 -> the output bond of the gate on q0
+            # i1 -> the input bond of the gate on q1
+            # o1 -> the output bond of the gate on q1
+            gate_bond = ["o0", "o1", "i0", "i1"]
+
+            aux_bonds = [f"q{x}" for x in range(n_qbonds)] + ["p"]
+            node_bonds = aux_bonds.copy()
+            node_bonds[bond_q0] = "i0"
+            node_bonds[bond_q1] = "i1"
+            result_bonds = aux_bonds.copy()

```

Comment:
You don't use `aux_bonds` anymore below, so perhaps no need to copy.

Reply:
Done